src/classes/RD2_OpportunityService.cls

```diff
@@ -237,23 +248,63 @@ public with sharing class RD2_OpportunityService {
                 currentDate, matcher.getNumberOfOpportunities(), rd.RecurringDonationSchedules__r
             ));
 
+            // Track unmatched matcher records so we can connect Opps and RDs from RDs with altered schedules
+            Boolean openOpportunityMatched = false;
+            Opportunity earliestUnmatchedOpportunity;
+            RD2_ScheduleService.Installment earliestUnmatchedInstallment;
+
             for (RD2_OpportunityMatcher.Record record : matcher.getRecords()) {
+
+                if (record.hasInstallment() && record.isNew()) {
+                    if (earliestUnmatchedInstallment == null ||
+                        earliestUnmatchedInstallment.nextDonationDate > record.getInstallment().nextDonationDate
+                    ) {
+                        earliestUnmatchedInstallment = record.getInstallment();
+                    }
+                }
+
                 if (record.isNew() || !record.isOpen()) {
                     continue;
                 }
 
                 Opportunity opp = record.getOpportunity();
                 if (record.hasInstallment()) {
                     installmentByOppId.put(opp.Id, record.getInstallment());
+                    openOpportunityMatched = true;
+                }
+
+                if (!record.hasInstallment()){
+                    if (earliestUnmatchedOpportunity == null ||
+                        earliestUnmatchedOpportunity.CloseDate > record.getOpportunity().CloseDate
+                    ) {
+                        earliestUnmatchedOpportunity = record.getOpportunity();
+                    }
                 }
 
-                Boolean isChanged = syncOppWithRecurringDonation(opp, rd);
-                isChanged = syncOppWithInstallment(opp, record.getInstallment()) || isChanged;
-                if (isChanged) {
+                Boolean wasOpportunityChanged = syncOppWithRecurringDonation(opp, rd);
+                wasOpportunityChanged = syncOppWithInstallment(opp, record.getInstallment()) || wasOpportunityChanged;
+                if (wasOpportunityChanged) {
                     opp.Name = buildOpportunityName(rd.npe03__Donor_Name__c, opp);
                     oppsToUpdate.add(opp);
                 }
             }
+
+            if (!openOpportunityMatched) {
+                if (earliestUnmatchedOpportunity == null || earliestUnmatchedInstallment == null) {
+                    break;
```

Comment:
Does this drop out of the loop entirely? Line 267 uses `continue;` rather than break.

Reply:
Yes, good catch.  I think this is an error.  I'll take this back and do some testing to make sure.